File: scripts/audit_jerga_parte.py
#!/usr/bin/env python3
"""¿QUE PALABRAS DE LA JERGA ESTAN PARTIENDO NOMBRES REALES? (2026-09-08)

# Por que existe

`INTEL_JARGON` cierra el nombre que se estuviera montando, y casi siempre acierta. Pero cuando la
palabra de jerga es **la segunda mitad de un nombre propio**, corta a la persona por el medio:

    «Iam Neutral»  ->  piloto «Iam»       (`neutral` esta en la jerga)
    «CCTV Eyes»    ->  piloto «CCTV»      (`eyes` esta en la jerga)

Es el patron de «Dee Yona», donde el que cortaba era un SISTEMA. Alli no se adivino: se proponen
las dos lecturas y decide `name_cache`. Aqui hara falta lo mismo — pero **primero se mide**, porque
el error del dia anterior fue el contrario: descartar el mecanismo de siglas mirando un solo caso
(`vni`) y dejar `eni` —487 avistamientos falsos— dos columnas mas abajo sin mirar.

# Que contesta este informe

  1. CUANTAS palabras de la jerga cortan nombres, no solo `neutral` y `eyes`.
  2. A QUIEN cortan, con las dos lecturas puestas una al lado de otra.
  3. QUE SE ARREGLA GRATIS: los nombres largos que `name_cache` YA confirma, sin pedir nada a ESI.
  4. QUE COSTARIA cerrarlo del todo: cuantos nombres distintos habria que preguntar, una vez.

⚠️ Solo LEE la base de datos (`mode=ro`) y enmascara los codigos de sistema de null antes de
   imprimir nada. Sin `--esi` no se manda ni una peticion.

⚠️ ESTE INFORME MIDE EL TROCEADOR DE HOY, no el de ayer. Desde que `intel.ts` sabe resolver la
   lectura larga, **los pares ya arreglados DESAPARECEN de aqui**: si «Iam Neutral» esta confirmado
   en `name_cache`, el piloto que sale ya es el largo y no hay corte que contar. O sea que las
   cifras bajando entre dos pasadas es la señal de que funciona, no de que se haya roto algo.
   Para ver el otro lado —que se recupero— esta `verificar_jerga_parte.py`.

Uso: python scripts/audit_jerga_parte.py
     python scripts/audit_jerga_parte.py --limite 200000
     python scripts/audit_jerga_parte.py --esi      (pregunta por los nombres largos desconocidos)
"""
import json
import os
import shutil
import subprocess
import sys
import tempfile
import sqlite3

AQUI = os.path.dirname(os.path.abspath(__file__))
RAIZ = os.path.dirname(AQUI)
sys.path.insert(0, AQUI)

# ★ La mascara de privacidad y la busqueda de la base de datos se REUTILIZAN de `audit_intel.py`.
#   No se copian: un fallo de enmascarado arreglado en un sitio tiene que quedar arreglado en los
#   dos, y ya paso una vez (la mascara solo tapaba MAYUSCULAS y el intel se escribe en minuscula).
from audit_intel import enmascarar, preguntar_esi, ruta_bd  # noqa: E402

LIMITE = 0
if "--limite" in sys.argv:
    LIMITE = int(sys.argv[sys.argv.index("--limite") + 1])
ESI = "--esi" in sys.argv


def main() -> int:
    ruta = ruta_bd()
    if not os.path.isfile(ruta):
        print(f"No encuentro la base de datos en:\n  {ruta}")
        return 1
    node = shutil.which("node")
    if not node:
        print("No encuentro `node` en el PATH. Hace falta para trocear con el parser de la app.")
        return 1

    db = sqlite3.connect(f"file:{ruta}?mode=ro", uri=True)
    n = db.execute("SELECT COUNT(*) FROM intel_line").fetchone()[0]
    if n == 0:
        print("`intel_line` esta vacia: importa el historico antes (Ajustes -> Intel).")
        return 1
    print(f"Base de datos: {ruta}")
    print(f"Lineas guardadas: {n:,}{f'  (se auditan {LIMITE:,})' if LIMITE else ''}\n")

    tmp = tempfile.mkdtemp(prefix="koru-jerga-")
    jsonl, cache_j, salida = (os.path.join(tmp, x) for x in ("lineas.jsonl", "cache.json", "out.json"))
    sql = "SELECT texto FROM intel_line" + (f" LIMIT {LIMITE}" if LIMITE else "")
    with open(jsonl, "w", encoding="utf-8") as f:
        for (texto,) in db.execute(sql):
            f.write(json.dumps({"texto": texto}, ensure_ascii=False) + "\n")

    cache = {}
    for nl, cid in db.execute("SELECT name_lower, character_id FROM name_cache"):
        if cid is not None:
            cache[nl] = 1 if cid > 0 else -1
    with open(cache_j, "w", encoding="utf-8") as f:
        json.dump(cache, f, ensure_ascii=False)
    print(f"  name_cache: {len(cache):,} nombres con veredicto de ESI")

    r = subprocess.run(
        [node, os.path.join(AQUI, "audit_jerga_parte.mjs"), RAIZ, jsonl, cache_j, salida],
        capture_output=True, text=True,
    )
    print(r.stdout.strip())
    if r.returncode != 0:
        print("Fallo al trocear:\n" + (r.stderr[-2000:] or "(sin salida)"))
        return 1
    with open(salida, encoding="utf-8") as f:
        d = json.load(f)
    shutil.rmtree(tmp, ignore_errors=True)

    print("\n" + "=" * 74)
    print("CUANTO PASA ESTO")
    print("=" * 74)
    print(f"  lineas auditadas                          {d['lineas']:>10,}")
    print(f"  lineas que sacan al menos un piloto       {d['lineasConPiloto']:>10,}")
    print(f"  lineas donde la jerga CORTA un nombre     {d['lineasCortadas']:>10,}"
          f"   ({100*d['lineasCortadas']/(d['lineasConPiloto'] or 1):.1f} % de las que sacan piloto)")
    print(f"  cortes por DETRAS  («Iam» + «Neutral»)    {d['totalDetras']:>10,}")
    print(f"  cortes por DELANTE («Red» + «Fox»)        {d['totalDelante']:>10,}")
    print(f"  TECHO: cortes donde hoy NO sale nadie     {d['totalPerdidos']:>10,}   <- el caso peor")
    print("     ^ la mitad corta tampoco llega a piloto («CCTV Eyes»: ESI ya dijo que CCTV no")
    print("       existe), asi que la persona se pierde ENTERA y en silencio. Es un TECHO, no un")
    print("       hecho: aqui no hay veredicto del troceador, solo que ningun catalogo la reclama.")
    print(f"\n  de todo eso, el nombre largo YA confirmado por ESI: {d['confirmadosYa']:,}")
    print("     ^ eso es lo que se recupera SIN gastar una sola peticion.")

    for clave, etq, ejemplo in [
        ("jergaDetras", "LA JERGA QUE CORTA POR DETRAS", "«Iam» + «Neutral» -> «Iam Neutral»"),
        ("jergaDelante", "LA JERGA QUE CORTA POR DELANTE", "«Red» + «Fox» -> «Red Fox»"),
        ("jergaPerdidos", "LA JERGA QUE HACE DESAPARECER A ALGUIEN (techo)", "«CCTV» + «Eyes»"),
    ]:
        filas = d.get(clave) or []
        if not filas:
            continue
        print("\n" + "=" * 74)
        print(f"{etq}   ({ejemplo})")
        print("  veces = cortes totales   gente = personas distintas cortadas")
        print("  si/no/? = veredicto de name_cache sobre el nombre LARGO (?  = nadie lo ha preguntado)")
        print("=" * 74)
        print("  MAY   = veces que la palabra venia con mayuscula inicial (pista, NO veredicto)")
        print("  ratio = veces / gente  <-- LA PUERTA. Un apellido lo lleva poca gente y se repite")
        print("          mucho; una palabra de jerga la lleva todo el mundo detras.")
        print("  obj   = la lectura larga es el principio del nombre de un OBJETO del mercado")
        print(f"  {'palabra':<14} {'ratio':>7} {'veces':>8} {'MAY':>7} {'gente':>7} "
              f"{'si':>7} {'no':>7} {'?':>8} {'obj':>7}")
        # Se imprimen TODAS las que pasan la puerta y las 15 siguientes: el corte hay que VERLO.
        # El resto es cola larga de jerga confirmada y solo se cuenta, para no llenar la pantalla.
        arriba = [x for x in filas if x["ratio"] >= d["ratio"]]
        abajo = [x for x in filas if x["ratio"] < d["ratio"]]
        for x in arriba:
            print(f"  {enmascarar(x['jerga']):<14} {x['ratio']:>7} {x['veces']:>8,} "
                  f"{x['mayus']:>7,} {x['distintos']:>7,} {x['si']:>7,} {x['no']:>7,} "
                  f"{x['duda']:>8,} {x['obj']:>7,}")
        print(f"  {'':-<14}-- PUERTA: ratio {d['ratio']} --{'':-<40}")
        for x in abajo[:15]:
            print(f"  {enmascarar(x['jerga']):<14} {x['ratio']:>7} {x['veces']:>8,} "
                  f"{x['mayus']:>7,} {x['distintos']:>7,} {x['si']:>7,} {x['no']:>7,} "
                  f"{x['duda']:>8,} {x['obj']:>7,}")
        if len(abajo) > 15:
            print(f"  ... y {len(abajo)-15} palabras mas, todas por debajo de "
                  f"{abajo[15]['ratio']} — es la jerga de siempre (GRUPO DE CONTROL).")

    cp = d.get("candidatosPuerta") or []
    print("\n" + "=" * 74)
    print(f"★ LO QUE PASA LA PUERTA (ratio >= {d['ratio']})  —  ESTA es la lista que hay que juzgar")
    print("=" * 74)
    print(f"  cortes que pasan la puerta: {d['cortesPuerta']:,}  "
          f"(de ellos {d['cortesPuertaObjeto']:,} son un OBJETO del mercado, no una persona)")
    print(f"  {'veces':>7} {'MAY':>5}  {'lectura corta':<24} {'lectura larga':<30} corto largo obj")
    for x in cp[:70]:
        print(f"  {x['veces']:>7,} {x['mayus']:>5,}  {enmascarar(x['corto']):<24} "
              f"{enmascarar(x['largo']):<30} {x['esiCorto']:>5} {x['esiLargo']:>5} "
              f"{'OBJ' if x['obj'] else '':>4}")
    if len(cp) > 70:
        print(f"  ... y {len(cp)-70} pares mas.")

    print("\n" + "=" * 74)
    print("A QUIEN CORTA  (todo, sin puerta — las dos lecturas, sin elegir)")
    print("  esi corto / esi largo:  si = ESI lo confirma · no = ESI dice que no existe · ? = sin preguntar")
    print("  ⚠️ 'largo = si' NO cierra el caso por si solo: existe gente llamada «Small» o «ESS».")
    print("     Lo que hace probable la lectura larga es que la corta NO exista, o que la larga se")
    print("     escriba asi una y otra vez en el mismo canal.")
    print("=" * 74)
    print(f"  {'veces':>7} {'MAY':>5}  {'lectura corta':<24} {'lectura larga':<30} corto largo")
    for x in (d.get("candidatosDetras") or [])[:60]:
        print(f"  {x['veces']:>7,} {x['mayus']:>5,}  {enmascarar(x['corto']):<24} "
              f"{enmascarar(x['largo']):<30} {x['esiCorto']:>5} {x['esiLargo']:>5}")

    for clave, etq in [("candidatosDelante", "por delante"),
                       ("candidatosPerdidos", "los que hoy NO sacan a nadie (techo)")]:
        cd = d.get(clave) or []
        if cd:
            print(f"\n  --- {etq} ---")
            for x in cd[:30]:
                print(f"  {x['veces']:>7,} {x['mayus']:>5,}  {enmascarar(x['corto']):<24} "
                      f"{enmascarar(x['largo']):<30} {x['esiCorto']:>5} {x['esiLargo']:>5}")

    dudas = d.get("dudas") or []
    dudas_may = d.get("dudasMayus") or []
    print("\n" + "=" * 74)
    print("LO QUE COSTARIA CERRARLO DEL TODO")
    print("=" * 74)
    dudas_p = d.get("dudasPuerta") or []
    print(f"  Preguntando por TODO:                  {len(dudas):>7,} nombres, "
          f"{(len(dudas) + 249) // 250:>4} peticiones   <- impagable")
    print(f"  Solo los que venian CON MAYUSCULA:     {len(dudas_may):>7,} nombres, "
          f"{(len(dudas_may) + 249) // 250:>4} peticiones")
    print(f"  ★ Solo los que PASAN LA PUERTA:        {len(dudas_p):>7,} nombres, "
          f"{(len(dudas_p) + 249) // 250:>4} peticiones")
    print("     (y sin los que ya son un objeto del mercado: eso es catalogo, no ESI)")
    print("  (`/universe/ids/` acepta 250 nombres por peticion; la respuesta se guarda para siempre)")
    print("  ⚠️ La mayuscula NO es prueba de nada —eso quedo desmentido hoy— pero SI es un buen")
    print("     filtro para no preguntar por «Fulano Perez on»: aqui solo decide a quien se")
    print("     PREGUNTA, nunca quien es una persona. Eso lo sigue contestando ESI.")
    # ⚠️ `--esi` pregunta SOLO por lo que pasa la puerta, nunca por los 85.903. Los limites de ESI
    #    cuentan PETICIONES desde diciembre de 2025, y 344 peticiones por una auditoria no se pagan.
    #    Si algun dia hiciera falta preguntar por todo, que sea una decision suya y explicita.
    dudas = dudas_p
    if dudas and not ESI:
        print("\n  Para preguntarlo ahora:  python scripts/audit_jerga_parte.py --esi")
        print(f"  (solo los {len(dudas):,} que pasan la puerta, "
              f"{(len(dudas) + 249) // 250} peticiones — nunca los {len(d.get('dudas') or []):,})")
    elif dudas and ESI:
        print(f"\n  Preguntando a ESI por {len(dudas):,} nombres (los que pasan la puerta)...")
        reales = preguntar_esi(dudas)
        print(f"  ESI reconoce como PERSONAJE {len(reales):,} de {len(dudas):,}:\n")
        # Se ordenan por las veces que aparecian, no alfabeticamente: lo primero que se lee tiene
        # que ser lo que mas dano hace hoy.
        veces = {}
        for k in ("candidatosDetras", "candidatosDelante", "candidatosPerdidos"):
            for x in (d.get(k) or []):
                veces[x["largo"].lower()] = x["veces"]
        enc = sorted(reales.items(), key=lambda kv: -veces.get(kv[0], 0))
        for i, (lc, canonico) in enumerate(enc[:80]):
            print(f"  {veces.get(lc, 0):>7,}  {enmascarar(canonico):<26}",
                  end="\n" if i % 2 == 1 else "")
        print()
        print(f"\n  Suman {sum(veces.get(lc, 0) for lc in reales):,} cortes a personas REALES.")
        print("  ⚠️ Esto NO se ha escrito en `name_cache`: este script solo lee. La respuesta la")
        print("     guarda la app cuando el troceador proponga las dos lecturas.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/audit_jerga_parte.mjs
// ★★ ¿QUÉ PALABRAS DE LA JERGA ESTÁN PARTIENDO NOMBRES REALES? (2026-09-08)
//
// # Por qué existe
//
// La lista `INTEL_JARGON` cierra el nombre que se estuviera montando. Es lo correcto casi siempre
// («Fulano on the gate» → el piloto es Fulano), pero cuando la palabra de jerga **es la segunda
// mitad de un nombre propio**, corta a la persona por el medio y ficha solo su primera palabra:
//
//     «Iam Neutral»  →  piloto «Iam»          (`neutral` está en la jerga)
//     «CCTV Eyes»    →  piloto «CCTV»         (`eyes` está en la jerga)
//
// Es exactamente el patrón de «Dee Yona», donde el que cortaba era un SISTEMA. Allí no se adivinó:
// se proponen las dos lecturas (`pilotAlts`) y decide `name_cache`. Aquí hará falta lo mismo — pero
// **primero hay que saber cuántas palabras hacen esto y a cuánta gente**, porque el error del día
// anterior fue justo el contrario: generalizar desde un solo caso (`vni`) y dejar `eni` —487
// avistamientos falsos— dos columnas más abajo sin mirar.
//
// # La regla que lo hace fiable
//
// ⚠️ ESTO **NO TIENE TROCEADOR PROPIO**, igual que `audit_intel.mjs`. Los pilotos los saca
//    `classifyIntel` de `src/intel.ts` —el troceador de la app— y aquí solo se mira **qué token va
//    justo detrás (o justo delante) del nombre que Koru ha extraído de verdad**. Si esto
//    reclasificara por su cuenta, mediría al auditor y no a Koru.
//
// ⚠️ Solo LEE. No decide nada, no toca `intel.ts` y no gasta una sola petición a ESI (eso lo hace
//    el Python, y solo si se le pide con `--esi`).
//
// Lo alimenta `audit_jerga_parte.py`, que es quien sabe leer SQLite.
import { readFileSync, writeFileSync, createReadStream } from "node:fs";
import { createInterface } from "node:readline";
import { join } from "node:path";
import { pathToFileURL } from "node:url";

const [, , RAIZ, JSONL, CACHE, SALIDA] = process.argv;
const ruta = (...p) => join(RAIZ, ...p);
// `import()` exige una URL: en Windows `C:\...` no lo es. Misma trampa que ya pisamos en audit_intel.
const ts = (await import(pathToFileURL(ruta("node_modules", "typescript", "lib", "typescript.js")).href)).default;
const tp = (src) =>
  ts.transpileModule(src, {
    compilerOptions: { module: ts.ModuleKind.ESNext, target: ts.ScriptTarget.ES2022 },
  }).outputText;
const M = await import(
  "data:text/javascript;base64," +
    Buffer.from(tp(readFileSync(ruta("src", "intel.ts"), "utf8"))).toString("base64")
);

// --- Los mismos catálogos que usa la app: el veredicto tiene que ser el suyo ---
const ne = JSON.parse(readFileSync(ruta("public", "neweden.json"), "utf8"));
const nameIdx = new Map(ne.systems.map((s) => [s.n.toLowerCase(), s]));
const zonaIdx = M.zonasDe(ne);
const shipNames = new Map([
  ...Object.entries(JSON.parse(readFileSync(ruta("public", "ship_names_i18n.json"), "utf8"))),
  ...Object.entries(JSON.parse(readFileSync(ruta("public", "ship_names.json"), "utf8"))),
]);
/** ★ EL CATÁLOGO DE OBJETOS, para separar dos problemas que parecían uno (2026-09-08).
 *
 *  «Sisters Combat Scanner Probe» ficha DOS pilotos falsos (`Sisters`, `Scanner`) y encima una nave
 *  falsa (la fragata Probe). Comprobado ejecutando el troceador, no razonado. Pero eso **no lo
 *  arregla `name_cache`**: no hay ninguna persona que confirmar, hay un OBJETO que el troceador no
 *  conoce porque `ship_names` solo trae naves. Si no se separan, el informe pide preguntarle a ESI
 *  por cosas que ya están en un fichero del disco — justo lo contrario de «catálogo primero».
 *
 *  Se indexan los prefijos de palabra de los 19.369 tipos del mercado, igual que `prefijosDe` hace
 *  con las naves: así «Combat Scanner» casa con «Combat Scanner Probe I». Es una PISTA para el
 *  informe, no un veredicto. */
const prefijosObjeto = new Set();
try {
  for (const t of JSON.parse(readFileSync(ruta("public", "market_types.json"), "utf8"))) {
    const partes = (t.n || "").toLowerCase().split(" ");
    for (let k = 1; k <= partes.length; k++) prefijosObjeto.add(partes.slice(0, k).join(" "));
  }
} catch { /* sin catálogo de objetos el informe sigue valiendo, solo pierde una columna */ }

const cache = JSON.parse(readFileSync(CACHE, "utf8"));
const noExisten = new Set(Object.entries(cache).filter(([, v]) => v === -1).map(([k]) => k));
const existen = new Set(Object.entries(cache).filter(([, v]) => v === 1).map(([k]) => k));

// ★ La jerga se lee del PROPIO `intel.ts`, no de una copia. Si mañana se añade una palabra, este
//   informe la mide sin tocarlo. Si cambiara el formato del fichero, se planta en vez de mentir.
const fuente = readFileSync(ruta("src", "intel.ts"), "utf8");
const conjunto = (nombre) => {
  const m = fuente.match(new RegExp(`const ${nombre} = new Set\\(\\[([\\s\\S]*?)\\]\\)`));
  if (!m) throw new Error(`No encuentro ${nombre} en intel.ts — ¿cambió el formato?`);
  return new Set([...m[1].matchAll(/"((?:[^"\\]|\\.)*)"/g)].map((x) => x[1].toLowerCase()));
};
const JERGA = conjunto("INTEL_JARGON");
// Las partículas («the», «van», «de») ya tienen su propio mecanismo dentro del troceador, así que
// aquí se excluyen: contarlas sería medir dos veces un problema que ya está resuelto.
const PARTICULAS = conjunto("PARTICULAS_NOMBRE");
console.log(`  jerga leída de intel.ts: ${JERGA.size} palabras`);

// El MISMO recorte de puntuación que hace `classifyIntel` (`clean`). Está copiado a propósito y es
// lo único que se copia: sin él, «Neutral,» no casaría con `neutral` y el informe diría que no pasa
// nada. Si algún día `clean` cambia en intel.ts, hay que cambiarlo aquí.
const clean = (s) => s.replace(/[*.,;:!?()]+$/g, "").replace(/^[*([]+/g, "").trim();

/** corto + jerga → { veces, mayus, corto, jerga, orig } */
const detras = new Map();
const delante = new Map();
/** ★ EL CASO QUE MI PRIMERA VERSIÓN NO VEÍA, y lo enseñó EJECUTARLA con «CCTV Eyes» de prueba.
 *
 *  Este informe empezó mirando solo a los pilotos que Koru SÍ saca, para medir a Koru y no al
 *  auditor. Pero cuando la mitad corta **no llega a piloto** —porque ESI ya dijo que «CCTV» no es
 *  nadie, o porque se queda corta— no hay piloto que mirar y el corte se vuelve INVISIBLE: la línea
 *  no ficha a nadie equivocado, simplemente **pierde a la persona entera y en silencio**. Es el caso
 *  peor, y era justo el que se me escapaba.
 *
 *  ⚠️ Aquí no hay veredicto del troceador, así que esto es un TECHO, no un hecho: se exige que
 *  ningún catálogo reclame esa palabra, pero no se comprueban abreviaturas ni apodos. Se imprime
 *  aparte y etiquetado — mezclarlo con lo medido de verdad sería inflar la cifra que justifica el
 *  trabajo. */
const perdidos = new Map();
let lineas = 0;
let lineasConPiloto = 0;
let lineasCortadas = 0;

const anota = (mapa, corto, jerga, jergaOriginal) => {
  const k = `${corto.toLowerCase()}|${jerga}`;
  let e = mapa.get(k);
  if (!e) mapa.set(k, (e = { veces: 0, mayus: 0, corto, jerga, orig: jerga }));
  e.veces++;
  // ¿Se escribió la palabra de jerga con mayúscula inicial? **No decide nada** —ESI demostró que hay
  // nombres canónicos en minúscula— pero dice si quien escribía la trataba como parte de un nombre,
  // y es lo que separa «Iam Neutral» de «Fulano on the gate». Se CUENTA, no se filtra: filtrar aquí
  // sería volver a meter en el medidor el axioma que este mismo día resultó falso.
  if (/^\p{Lu}/u.test(jergaOriginal)) {
    e.mayus++;
    e.orig = jergaOriginal;
  }
};
/** ¿Reclama algún catálogo esta palabra? Cuatro consultas, no una clasificación. */
const loReclamaElCatalogo = (lc) =>
  JERGA.has(lc) || PARTICULAS.has(lc) || nameIdx.has(lc) || shipNames.has(lc) || zonaIdx.has(lc);
/** Forma mínima de algo que pueda ser un trozo de nombre: letras y al menos dos. */
const FORMA_TROZO = /^[\p{L}][\p{L}'-]+$/u;

const rl = createInterface({ input: createReadStream(JSONL, "utf8"), crlfDelay: Infinity });
for await (const linea of rl) {
  if (!linea.trim()) continue;
  const { texto } = JSON.parse(linea);
  lineas++;
  const p = M.classifyIntel(texto, nameIdx, shipNames, noExisten, zonaIdx, existen);
  if (p.pilots.length) lineasConPiloto++;
  // Los campos, con la MISMA limpieza y el MISMO corte que el troceador.
  const campos = M.limpiarMarcadoEve(texto).split(/\s{2,}/).map((f) => f.trim()).filter(Boolean);
  let cortada = false;
  for (const campo of campos) {
    const tokens = campo.split(/\s+/).map(clean).filter(Boolean);
    const bajos = tokens.map((t) => t.toLowerCase());
    /** Posiciones ya explicadas por un piloto de verdad: lo demás es el bloque «techo». */
    const cubierto = new Set();
    for (const piloto of p.pilots) {
      const partes = piloto.split(/\s+/).map((t) => t.toLowerCase());
      for (let i = 0; i + partes.length <= tokens.length; i++) {
        let casa = true;
        for (let j = 0; j < partes.length; j++)
          if (bajos[i + j] !== partes[j]) { casa = false; break; }
        if (!casa) continue;
        for (let j = 0; j < partes.length; j++) cubierto.add(i + j);
        // ★ DETRÁS: «Iam» seguido de «Neutral». La lectura larga sería «Iam Neutral».
        const sig = i + partes.length;
        if (sig < tokens.length && JERGA.has(bajos[sig]) && !PARTICULAS.has(bajos[sig])) {
          anota(detras, piloto, bajos[sig], tokens[sig]);
          cortada = true;
        }
        // ★ DELANTE: jerga justo antes del nombre («Red» + «Fox»). El mismo mecanismo, al revés.
        if (i > 0 && JERGA.has(bajos[i - 1]) && !PARTICULAS.has(bajos[i - 1])) {
          anota(delante, piloto, bajos[i - 1], tokens[i - 1]);
          cortada = true;
        }
      }
    }
    // ★ EL TECHO: jerga precedida de algo con forma de nombre que HOY no sale como piloto.
    for (let i = 1; i < tokens.length; i++) {
      if (!JERGA.has(bajos[i]) || PARTICULAS.has(bajos[i])) continue;
      if (cubierto.has(i - 1)) continue;
      const prev = bajos[i - 1];
      if (!FORMA_TROZO.test(prev) || loReclamaElCatalogo(prev)) continue;
      // ⚠️ FALSO POSITIVO MÍO, cazado ejecutando el troceador y no leyéndolo: «Caldari Shuttle» SÍ
      //    está en `ship_names` y Koru la reconoce perfectamente. Mi bucle miraba la palabra suelta
      //    (`caldari` no es una nave) y contaba 609 cortes que no existen. La comprobación tiene
      //    que ir sobre el PAR, que es lo que el troceador mira de verdad (`naveDesde` prueba de
      //    más largo a más corto). Se miran también tres palabras: «Council Diplomatic Shuttle».
      if (shipNames.has(`${prev} ${bajos[i]}`)) continue;
      if (i >= 2 && shipNames.has(`${bajos[i - 2]} ${prev} ${bajos[i]}`)) continue;
      anota(perdidos, tokens[i - 1], bajos[i], tokens[i]);
    }
  }
  if (cortada) lineasCortadas++;
}

/** El veredicto que ya tenemos apuntado, sin preguntar nada. */
const veredicto = (nombre) => {
  const v = cache[nombre.toLowerCase()];
  return v === 1 ? "si" : v === -1 ? "no" : "?";
};

const filas = (mapa, alReves) =>
  [...mapa.values()]
    .map((e) => {
      const largo = alReves ? `${e.orig} ${e.corto}` : `${e.corto} ${e.orig}`;
      return {
        corto: e.corto,
        jerga: e.jerga,
        largo,
        veces: e.veces,
        mayus: e.mayus,
        esiCorto: veredicto(e.corto),
        esiLargo: veredicto(largo),
        // ¿La lectura larga es el principio del nombre de un OBJETO del mercado? Entonces esto no
        // es una persona que preguntar, es un catálogo que no estamos mirando.
        obj: prefijosObjeto.has(largo.toLowerCase()),
      };
    })
    .sort((a, b) => b.veces - a.veces);

const fd = filas(detras, false);
const fl = filas(delante, true);
const fp = filas(perdidos, false);

/** ★★ LA PUERTA: VECES ÷ GENTE. Salió de mirar por qué `nv` encabezaba la lista y no debía.
 *
 *  `nv` corta 38.440 veces… repartidas entre **17.618 personas distintas**. Eso no es un apellido:
 *  es que en su intel `nv` significa «NUEVO» —que los hostiles han vuelto a aparecer, aunque el
 *  aviso sea idéntico al de hace un minuto— y por eso lo lleva detrás medio universo. Palabras suyas
 *  (2026-09-08), y ningún catálogo podía contestarlo.
 *
 *  La señal es la CONCENTRACIÓN: un apellido lo lleva poca gente y se repite mucho.
 *
 *      nv        38.440 / 17.618 =   2,2      in    2.814 / 2.057 = 1,4   ← jerga
 *      eyes       3.208 /     88 =  36,5      neutral 1.248 /  13 =  96   ← trozo de nombre
 *      meme       1.440 /     10 = 144
 *
 *  ⚠️ Y tiene GRUPO DE CONTROL gratis, que es lo que le faltó a la dispersión de ayer: las ~198
 *  palabras de `INTEL_JARGON` son todas jerga confirmada, así que si la señal sirve, casi todas
 *  tienen que caer en la banda baja. Por eso el informe imprime la lista ENTERA ordenada por este
 *  número: el umbral se ve, no se cree. */
const RATIO = Number(process.env.KORU_RATIO ?? 10);
/** Resumen por palabra de jerga: a cuánta gente distinta corta y con qué veredicto. */
const resumenJerga = (fs) => {
  const m = new Map();
  for (const f of fs) {
    let e = m.get(f.jerga);
    if (!e) m.set(f.jerga, (e = { jerga: f.jerga, veces: 0, mayus: 0, distintos: 0, si: 0, no: 0, duda: 0, obj: 0 }));
    e.veces += f.veces;
    e.mayus += f.mayus;
    e.distintos++;
    if (f.obj) e.obj += f.veces;
    if (f.esiLargo === "si") e.si += f.veces;
    else if (f.esiLargo === "no") e.no += f.veces;
    else e.duda += f.veces;
  }
  return [...m.values()]
    .map((e) => ({ ...e, ratio: +(e.veces / e.distintos).toFixed(1) }))
    .sort((a, b) => b.ratio - a.ratio);
};
const rd = resumenJerga(fd);
const rl2 = resumenJerga(fl);
const rp = resumenJerga(fp);
/** Las palabras que pasan la puerta, por dirección. */
const pasan = (res) => new Set(res.filter((x) => x.ratio >= RATIO).map((x) => x.jerga));
const pasaD = pasan(rd), pasaL = pasan(rl2), pasaP = pasan(rp);
const conPuerta = [
  ...fd.filter((f) => pasaD.has(f.jerga)),
  ...fl.filter((f) => pasaL.has(f.jerga)),
  ...fp.filter((f) => pasaP.has(f.jerga)),
];

// Las preguntas que costaría cerrarlo. Se dan las tres cifras porque son tres decisiones distintas:
// preguntar por todo, por lo que venía con mayúscula, o solo por lo que pasa la puerta.
const todas = [...fd, ...fl, ...fp];
const dudas = [...new Set(todas.filter((f) => f.esiLargo === "?").map((f) => f.largo))];
const dudasMayus = [
  ...new Set(todas.filter((f) => f.esiLargo === "?" && f.mayus > 0).map((f) => f.largo)),
];
// ⚠️ Los que ya son un objeto del mercado NO se preguntan: eso es catálogo, no ESI.
const dudasPuerta = [
  ...new Set(conPuerta.filter((f) => f.esiLargo === "?" && !f.obj).map((f) => f.largo)),
];

writeFileSync(
  SALIDA,
  JSON.stringify(
    {
      lineas,
      lineasConPiloto,
      lineasCortadas,
      ratio: RATIO,
      jergaDetras: rd,
      jergaDelante: rl2,
      jergaPerdidos: rp,
      candidatosDetras: fd.slice(0, 200),
      candidatosDelante: fl.slice(0, 120),
      candidatosPerdidos: fp.slice(0, 200),
      // Todo lo que pasa la puerta, sin recortar: es la lista que hay que juzgar.
      candidatosPuerta: conPuerta.sort((a, b) => b.veces - a.veces),
      cortesPuerta: conPuerta.reduce((a, b) => a + b.veces, 0),
      cortesPuertaObjeto: conPuerta.filter((f) => f.obj).reduce((a, b) => a + b.veces, 0),
      dudasPuerta,
      totalDetras: fd.reduce((a, b) => a + b.veces, 0),
      totalDelante: fl.reduce((a, b) => a + b.veces, 0),
      totalPerdidos: fp.reduce((a, b) => a + b.veces, 0),
      confirmadosYa: todas.filter((f) => f.esiLargo === "si").reduce((a, b) => a + b.veces, 0),
      dudas,
      dudasMayus,
    },
    null,
    1
  ),
  "utf8"
);
console.log(
  `  ${lineas.toLocaleString("es-ES")} líneas · ${lineasCortadas.toLocaleString("es-ES")} con un piloto cortado por jerga`
);
